net/socket.h

```diff
@@ -21,81 +21,217 @@ limitations under the License.
+            addr._in_addr_field[0] = nl1;
+            addr._in_addr_field[1] = nl2;
+            addr._in_addr_field[2] = nl3;
+            addr._in_addr_field[3] = nl4;
+        }
+        // V4 constructor (Internet Address)
+        explicit IPAddr(in_addr internet_addr) {
+            map_v4(internet_addr);
+        }
+        // V4 constructor (Network byte order)
+        explicit IPAddr(uint32_t nl) {
+            map_v4(nl);
+        }
+        // String constructor
+        explicit IPAddr(const char* s) {
+            if (inet_pton(AF_INET6, s, &addr) > 0) {
+                return;
+            }
+            in_addr v4_addr;
+            if (inet_pton(AF_INET, s, &v4_addr) > 0) {
+                map_v4(v4_addr);
+                return;
+            }
+            // Invalid string, make it a default value
+            *this = IPAddr();
+        }
+        // Check if it's actually an IPv4 address mapped in IPV6
+        bool is_ipv4() const {
+            if (ntohl(addr._in_addr_field[2]) != 0x0000ffff) {
+                return false;
+            }
+            if (addr._in_addr_field[0] != 0 || addr._in_addr_field[1] != 0) {
+                return false;
+            }
+            return true;
+        }
+        // We regard the default IPv4 0.0.0.0 as undefined
+        bool undefined() const {
+            return *this == V4Any();
+        }
+        // Should ONLY be used for IPv4 address
+        uint32_t to_nl() const {
+            return addr._in_addr_field[3];
+        }
+        bool is_loopback() const {
+            return is_ipv4() ? (*this == V4Loopback()) : (*this == V6Loopback());
+        }
+        bool is_broadcast() const {
+            // IPv6 does not support broadcast
+            return is_ipv4() && (*this == V4Broadcast());
+        }
+        bool is_link_local() const {
+            if (is_ipv4()) {
+                return (to_nl() & htonl(0xffff0000)) == htonl(0xa9fe0000);
+            } else {
+                return (addr._in_addr_field[0] & htonl(0xffc00000)) == htonl(0xfe800000);
+            }
+        }
+        bool operator==(const IPAddr& rhs) const {
+            return memcmp(this, &rhs, sizeof(rhs)) == 0;
+        }
+        bool operator!=(const IPAddr& rhs) const {
+            return !(*this == rhs);
+        }
+        std::string to_string() const {
+            std::string str;
+            char text[INET6_ADDRSTRLEN];
+            if (is_ipv4()) {
+                in_addr ip4;
+                ip4.s_addr = to_nl();
+                inet_ntop(AF_INET, &ip4, text, INET_ADDRSTRLEN);
+            } else {
+                inet_ntop(AF_INET6, &addr, text, INET6_ADDRSTRLEN);
+            }
+            str.assign(text, strlen(text));
+            return str;
+        }
+    public:
+        static IPAddr V6None() {
+            return IPAddr(htonl(0xffffffff), htonl(0xffffffff), htonl(0xffffffff), htonl(0xffffffff));
+        }
```

Comment:
Even though htonl(0xffffffff) is still 0xffffffff, add this to remind reader that the ip is stored in network byte order.